ehrapy/tools/nlp/_medcat.py

```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
+        res.to_csv(".")
+
+    overview_table = _df_to_rich_table(res.nlargest(n, "n_patients"))
+    console = Console()
+    console.print(overview_table)
+
+
+def _add_binary_column_to_obs(ep_cat: MedCAT, adata: AnnData, name: str) -> None:
+    """Adds a binary column to obs (temporarily) for plotting infos extracted from freetext.
+    Indicates whether the specific entity to color by has been found in that row or not.
+    """
+    # only extract affirmed entities
+    df = _filter_df_by_status(ep_cat.annotated_results, "Affirmed")
+    # currently, only the pretty_name column is supported
+    # TODO: add support for CUI/types and hierarchical entities (in case of diseases implement this with ICD mapping feature)
+    adata.obs[name] = df.groupby("row_nr").agg({"pretty_name": (lambda x: any(x.isin([name])))})
+
+
+def _annotated_results_to_df(flattened_results: dict) -> pd.DataFrame:
+    """Turn the flattened annotated results into a pandas DataFrame and remove duplicates."""
+    df = pd.DataFrame.from_dict(flattened_results, orient="index")
+    # remove duplicate entries; for example when a single entity like a disease is mentioned multiple times without any meaningful context changes
+    # Example: The patient suffers from Diabetes. Cause of the Diabetes, he receives drug X.
+    df.drop_duplicates(subset=["cui", "row_nr", "meta_anns"])
+    return df
+
+
+def _flatten_annotated_results(annotation_results: dict) -> dict:
+    """Flattens the nested set (usually 5 level nested) of annotation results.
+    annotation_results is just a simple flattened dict with infos on all entities found
+    """
+    flattened_annotated_dict = {}
+    entry_nr = 0
+
+    # row numbers where the text column is located in the original data
+    for row_id in annotation_results.keys():
+        # all entities extracted from a given row
+        entities = annotation_results[row_id]["entities"]
+        # iterate over all entities extracted from the specific row
+        for entity_id in entities.keys():
+            # ignore tokens for now
+            if entity_id != "tokens":
+                single_entity = {"row_nr": row_id}
+                entity = entities[entity_id]
+                # iterate over all info attributes of a single entity found in a specific row
+                for entity_key in entity.keys():
+                    if entity_key in ["pretty_name", "cui", "type_ids", "types"]:
+                        single_entity[entity_key] = entities[entity_id][entity_key]
+                    elif entity_key == "meta_anns":
+                        single_entity[entity_key] = entities[entity_id][entity_key]["Status"]["value"]
+                flattened_annotated_dict[entry_nr] = single_entity
+                entry_nr += 1
+    return flattened_annotated_dict
+
+
+def _format_df_column(df: pd.DataFrame, column_name: str) -> list[tuple[int, str]]:
+    """Format the df to match: formatted_data = [(row_id, row_text), (row_id, row_text), ...]
```

Comment:
comment style...

Maybe I have to turn on the docstring linting to annoy you haha :)

Reply:
we should haha